components/server/src/user/user-controller.ts

```diff
@@ -601,15 +601,22 @@ export class UserController {
     }
 
     protected getSafeReturnToParam(req: express.Request) {
-        const returnToURL: string | undefined = req.query.redirect || req.query.returnTo;
-        if (returnToURL) {
-            const hostUrl = this.env.hostUrl.url as URL;
-            if (returnToURL.toLowerCase().startsWith(`${hostUrl.protocol}//${hostUrl.host}`.toLowerCase())) {
-                return returnToURL;
-            }
-            if (returnToURL.toLowerCase().startsWith(this.env.brandingConfig.homepage.toLowerCase())) {
-                return returnToURL;
-            }
+        const where: string | undefined = req.query.redirect || req.query.returnTo;
+        if (!where) {
+            log.debug({ sessionId: req.sessionID }, "Empty redirect URL");
+            return;
+        }
+        try {
+            const redirectUrl = new URL(where).toString();
+            const hostUrl = (this.env.hostUrl.url as URL).toString();
+            const brandingUrl = new URL(this.env.brandingConfig.homepage).toString();
+            if (redirectUrl === hostUrl || redirectUrl === brandingUrl) {
```

Comment:
Checking for equality isn't working in the use cases this is meant for. For instance, when switching accounts the browser agent is sent to `/logout?returnTo=<login-url-with-preset-provider>`. The equality check would allow to return to the main URL only.

Reply:
I clearly didn't know about this use case. Thanks!